distributed/tests/test_worker_client.py

```diff
@@ -301,24 +301,41 @@ def f():
     assert result
 
 
-@pytest.mark.xfail(
-    reason="Flaky due to https://github.com/dask/distributed/issues/5915"
+@pytest.mark.parametrize(
+    "scheduler_addr_fact",
+    [
+        lambda s: s.address,
+        pytest.param(
+            lambda s: "localhost:" + s.address.split(":")[-1],
+            marks=pytest.mark.xfail(
+                reason="Flaky due to https://github.com/dask/distributed/issues/5915"
+            ),
+        ),
+    ],
 )
 @gen_cluster()
-async def test_submit_different_names(s, a, b):
```

Comment:
Despite it's being marked flaky recently, this test has been solid before the changes proposed in this PR. This was achieved rather coincidentally since Clients were already added to the global default_client dict after initialization, before they were actually started. 
Running this test on `main` yields an exception pointing this out.

```
>   assert c_inner.status == "running"
E   AssertionError: assert 'newly-created' == 'running'
E     - running
E     + newly-created
```

What's happening on main is that

* As soon as the futures are deserialized on the worker, a
  `get_client("tcp://localhost:XXX")` is called in `Future.__setstate__`
* localhost is resolved as part of `get_client`
* since now the input address matches with the worker address, we call `Worker._get_client`
* `Worker._get_client` however does *not* match the default client (the one we created manually in the test) https://github.com/dask/distributed/blob/30f0b601ca98bbaed679ae22ce77b6ededb7ce1e/distributed/worker.py#L4296-L4306
* It will create a new client using the Worker RPC which always uses the resolved address https://github.com/dask/distributed/blob/30f0b601ca98bbaed679ae22ce77b6ededb7ce1e/distributed/worker.py#L4314-L4324
* This new client will be set to `Worker._client` *and* is the new
  `default_client`. `main` does only filter global clients for `!= closed` and
  not `running`
* Since the new one is the default client, the compute call will get the same
  one and it will always match addresses since it was created using the worker
  scheduler RPC
* This Client is initialized using `asynchronous=True` but is not awaited. Therefore, it is not actually up, yet, which is why we receive the above `newly-created` status

Diff to this branch:
- This branch requries a default client to be actually running. This leads us to
  initialize potentially many clients causing this confusion

An ideal world would...
- Detect the default client as the manually created one using localhost
- Recognize that the localhost client is talking to the same scheduler
- Not initialize any further Clients